pkg/login/social/connectors/org_role_mapper.go

```diff
@@ -0,0 +1,185 @@
+package connectors
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const MapperMatchAllOrgID = -1
+
+type OrgRoleMapper struct {
+	cfg        *setting.Cfg
+	logger     log.Logger
+	orgService org.Service
+}
+
+func ProvideOrgRoleMapper(cfg *setting.Cfg, orgService org.Service) *OrgRoleMapper {
+	return &OrgRoleMapper{
+		cfg:        cfg,
+		logger:     log.New("orgrole.mapper"),
+		orgService: orgService,
+	}
+}
+
+func (m *OrgRoleMapper) MapOrgRoles(ctx context.Context, externalOrgs []string, orgMappingSettings []string, directlyMappedRole org.RoleType) (map[int64]org.RoleType, error) {
+	orgMapping := m.splitOrgMappingSettings(ctx, orgMappingSettings)
+
+	userOrgRoles := getMappedOrgRoles(externalOrgs, orgMapping)
+
+	m.handleGlobalOrgMapping(userOrgRoles)
+
+	if len(userOrgRoles) == 0 {
```

Comment:
We could maybe add a `role_strict` bool to the org mapper, to fail if we didn't have any mapping and no direct mapped role?

Reply:
I considered that and decided not to add it (at least for now), because strict role assignment is a behaviour of the OAuth providers. I'd like to have a generic OrgRoleMapper, because in the future this might be reused by the SAML provider as well.

Reply:
Sorry, this is still a bit unclear to me, in the OAuth providers, how can we fail with `role_attribute_strict` given this would return the default mapping if no mapping was found?

Reply:
> Sorry, this is still a bit unclear to me, in the OAuth providers, how can we fail with role_attribute_strict given this would return the default mapping if no mapping was found?

Handling the `role_attribute_strict` setting should be the responsibility of the connector (it only applies to OAuth and JWT, not for SAML, that's why I don't want to include it in here). This is the part that handles `role_attribute_strict`: https://github.com/grafana/grafana/pull/77683/files#diff-352c498ba3516f310e434de756435c67f68046dd577f9456292d7aca1025fb63R303

Reply:
Ok, from the code you shared, I think we are interpreting `role_attribute_strict` differently. So we need to clarify that.

To me `role_attribute_strict` was created to prevent users that have not been mapped to a valid role from logging in. Therefore, I consider that if your `org_mapping` gives you a valid role, then you can log in. On the other hand, if your `org_mapping` maps you in no organizations or to an invalid role, then you should not log in.

If I read the linked code correctly, to you, `role_attribute_strict` means that you ignore `org_mapping` altogether.

Reply:
Yeah what @gamab said is true if `role_attribute_strict` is set we need to prevent users from signing in if no valid role was found https://github.com/grafana/grafana/blob/main/pkg/login/social/connectors/generic_oauth.go#L279C29-L284.

This could still be handled outside of this component

Reply:
If this is handled outside of this component, we should not return the default value in case no mapping has been found. Otherwise the caller won't be able to make the distinction between no mapping and mapped to default.

Reply:
Oh yeah you would need to pass the flag and handle it inside, you are correct

Reply:
> To me role_attribute_strict was created to prevent users that have not been mapped to a valid role from logging in. Therefore, I consider that if your org_mapping gives you a valid role, then you can log in. On the other hand, if your org_mapping maps you to no organization or to an invalid role, then you should not log in.

I implemented this, so the result will be an empty mapping when 
- getAllOrgs fail in global org mapping (related to your other comment: https://github.com/grafana/grafana/pull/86973#discussion_r1581049871)
- roleString = true and
    - directlyMappedRole is empty and orgMapping is empty
    - directlyMappedRole is empty and orgMapping contains at least 1 invalid mapping
    - directlyMappedRole is empty and orgMapping is not empty, but evaluated to empty org roles
    - directlyMappedRole is empty and at least one org name resolution fails or the org is non-existent (basically this means that the `orgMapping` is empty, because `ParseOrgMappingSettings` returns an empty org mapping if it cannot find an org by org_name)